ord_schema/validations.py

```diff
@@ -0,0 +1,428 @@
+        # NOTE(ccoley): I made the conscious decision to raise an error here,
+        # rather than assume that the message is valid. If a message does not
+        # require any message-level checks (not uncommon), then it should still
+        # be listed in the dictionary switch above withpass. This will force
+        # us to think about what is necessary if/when new messages are added.
+        raise NotImplementedError(f"Don't know how to validate {type(message)}")
+
+
+class ValidationWarning(Warning):
+    pass
+
+def return_message_if_valid(func):
+    def wrapper(message):
+        func(message)
+        return message
+    return wrapper
+
+def ensure_float_nonnegative(message, field):
+    if getattr(message, field) < 0:
+        raise ValueError(f'Field {field} of message '\
+            f'{type(message).DESCRIPTOR.name} must be non-negative')
+
+def ensure_float_range(message, field, min, max):
+    if getattr(message, field) < min or getattr(message, field) > max:
+        raise ValueError(f'Field {field} of message '\
+            f'{type(message).DESCRIPTOR.name} must be between {min} and {max}')
+
+def ensure_units_specified_if_value_defined(message):
+    if message.value: # can't distinguish 0 and unspecified anyway
+        if message.units == message.UNSPECIFIED:
+            raise ValueError(f'Unspecified units for {type(message)} with ' \
+                f'value defined ({message.value})')
+
+def ensure_details_specified_if_type_custom(message):
+    if message.type == message.CUSTOM:
+        if not message.details:
+            raise ValueError(f'Custom type defined for {type(message)}, ' \
+                'but details field is empty')
+
+@return_message_if_valid
+def ValidateReaction(message):
+    if len(message.inputs) == 0:
+        raise ValueError('Reactions should have at least 1 reaction input')
+    # TODO(ccoley) Should outcomes also have a minimum length?
+
+@return_message_if_valid
+def ValidateReactionIdentifier(message):
+    ensure_details_specified_if_type_custom(message)
+
+@return_message_if_valid
+def ValidateReactionInput(message):
+    if len(message.components) == 0:
+        raise ValueError('Reaction inputs must have at least one component')
+
+@return_message_if_valid
+def ValidateCompound(message):
+    if len(message.identifiers) == 0:
+        raise ValueError('Compounds must have at least one identifier')
+
+@return_message_if_valid
+def ValidateCompoundFeature(message):
+    pass
+
+@return_message_if_valid
+def ValidateCompoundPreparation(message):
+    ensure_details_specified_if_type_custom(message)
+
+@return_message_if_valid
+def ValidateCompoundIdentifier(message):
+    ensure_details_specified_if_type_custom(message)
+    # TODO(ccoley): Add identifier-specific validation, e.g., by using
+    # RDKit to try to parse SMILES, looking up NAMEs using online resolvers
+
+@return_message_if_valid
+def ValidateVessel(message):
+    if message.type == message.VesselType.CUSTOM:
```

Comment:
Please combine these `if` statements. (I won't comment on every one.)

Reply:
Is it worth combining even if it makes the single if statement multiline?